src/Control/Distributed/Process/Internal/Primitives.hs

```diff
@@ -917,9 +919,17 @@ unlinkPort sport = do
 unmonitor :: MonitorRef -> Process ()
 unmonitor ref = do
   unmonitorAsync ref
-  receiveWait [ matchIf (\(DidUnmonitor ref') -> ref' == ref)
-                        (\_ -> return ())
-              ]
+  receiveWait [ matchIf isDidUnmonitor ignore ]
+  receiveAllNow [ matchIf isProcessMonitorNotification ignore,
```

Comment:
Actually, it looks to me like there can be at most one monitor notification in the mailbox for a given `MonitorRef`. Therefore no loop would be necessary. It should suffice:
```Haskell
_ <- receiveTimeout 0 [ ... ]
```